src/ops/fuelup_self.rs

```diff
@@ -1,17 +1,60 @@
-use anyhow::Result;
+use anyhow::{bail, Result};
+use std::{fs, path::Path};
+use tempfile::tempdir_in;
+use tracing::{error, info};
 
 use crate::{
     component,
     download::{download_file_and_unpack, unpack_bins, DownloadCfg},
-    path::fuelup_bin_dir,
+    path::{fuelup_bin, fuelup_bin_dir},
 };
 
+pub fn attempt_install_self(download_cfg: DownloadCfg, dst: &Path) -> Result<()> {
+    download_file_and_unpack(&download_cfg, dst)?;
+    unpack_bins(dst, dst)?;
+
+    Ok(())
+}
+
 pub fn self_update() -> Result<()> {
     let download_cfg = DownloadCfg::new(component::FUELUP, None)?;
+
+    let fuelup_bin = fuelup_bin();
     let fuelup_bin_dir = fuelup_bin_dir();
+    let fuelup_backup = fuelup_bin_dir.join("fuelup-backup");
+
+    let fuelup_new_dir = tempdir_in(&fuelup_bin_dir)?;
+
+    if let Err(e) = attempt_install_self(download_cfg, fuelup_new_dir.path()) {
+        // Skip all other steps if downloading fails here.
+        // We do not need to handle failure, since this downloads to a tempdir.
+        bail!("Failed to install fuelup: {}", e);
+    };
+
+    if fuelup_bin.exists() {
+        // Make a backup of fuelup, fuelup-backup.
+        fs::rename(&fuelup_bin, &fuelup_backup).expect("Could not make a fuelup-backup");
+    };
+
+    info!(
+        "Copying {} to {}",
+        fuelup_new_dir.path().join("fuelup").display(),
```

Comment:
We should construct this path once and re-use it here and at L44, especially as the name contains a string literal. Do we have this name as a const anywhere?

Reply:
Hmm, we don't have it as a const and I didn't include this within `path.rs` as I believe this is only relevant within the context of `fuelup self update`. `fuelup_new_dir` is also a `TempDir` (which I opted to use since it automatically cleans up after itself). I will create the path right after execution of `tempdir_in()`.